### PR TITLE
feat(pending): :sparkles: Added pending support

### PR DESCRIPTION
**Pull request by Trantorian1**
# Pull Request type

- Feature

## What is the current behavior?

Deoxys currently does not support pending blocks.

## What is the new behavior?

Pending blocks are synchronized at set time interval in `l2.rs` once the node is fully synchronized. RPC methods supporting pending block calls then refer to those when answering request.

### Methods supporting `Pending` calls:

| Method | Status |
|-|-|
|`getBlockWithTxs`|:white_check_mark:  completed|
|`getBlockWithTxHashes`|:white_check_mark: completed|
|`getTransactionReceipt`|:white_check_mark: completed|
|`getStateUpdate`|:warning: pending support only|

## Does this introduce a breaking change?

No. Pending blocks are only saved in RAM, so Node db does not need to be updated.

### CODE REVIEW

File: crates/client/sync/src/l2.rs [hunk 1]
```diff
@@ -477,16 +486,40 @@ pub fn verify_l2<B: BlockT>(
...
+        // This type conversion is evil and should not be necessary
+        *STARKNET_PENDING_STATE_UPDATE.write().expect("Failed to aquire write lock on STARKNET_PENDING_STATE_UPDATE") =
+            Some(crate::convert::state_update(state_update));
+    }
 
     let mut highest_block_hash_and_number = STARKNET_HIGHEST_BLOCK_HASH_AND_NUMBER
         .write()
         .expect("Failed to acquire write lock on STARKNET_HIGHEST_BLOCK_HASH_AND_NUMBER");
-    *highest_block_hash_and_number = (hash, number);
+    *highest_block_hash_and_number = (hash_current, number);
```
Comment (jbcaron) [inline crates/client/sync/src/l2.rs]: refactor it like the other 2 RwLock, without using a mutable variable

Comment (Trantorian1) [reply crates/client/sync/src/l2.rs]: Will update for the final pr :+1:

File: crates/client/rpc/src/pending/get_transaction_receipt.rs [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+    C: HeaderBackend<B> + BlockBackend<B> + StorageProvider<B, BE> + 'static,
+    C: ProvideRuntimeApi<B>,
+    C::Api: StarknetRuntimeApi<B> + ConvertTransactionRuntimeApi<B>,
+    G: GenesisProvider + Send + Sync + 'static,
+    H: HasherT + Send + Sync + 'static,
+{
+    let execution_infos = server
+		.client
+		.runtime_api()
+		.re_execute_transactions(previous_block_hash, vec![], transactions.clone())
```
Comment (jbcaron) [inline crates/client/rpc/src/pending/get_transaction_receipt.rs]: you can modify the parameters like this to retrieve only the information from the last transaction
```rust
re_re_execute_transactions(previous_block_hash, <previous_tx>, <last_tx>)
```
and remove `pop()`

File: crates/client/rpc/src/pending/get_transaction_receipt.rs [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+		})?
+		.map_err(|e| {
+			log::error!("Failed to reexecute the transactions: {e:?}");
+			StarknetRpcApiError::InternalServerError
+		})?
+		.map_err(|_| {
+			log::error!("One of the transaction failed during it's reexecution");
+			StarknetRpcApiError::InternalServerError
+		})?
+		.pop() // get only the last transaction execution info
```
Comment (jbcaron) [inline crates/client/rpc/src/pending/get_transaction_receipt.rs]: remove` pop()`
*in reference to the comment above*

File: crates/client/sync/src/l2.rs (hunk 1 above)
Comment (Trantorian1) [reply crates/client/sync/src/l2.rs]: done!

**Review by antiyro — APPROVED** [review]

File: crates/client/rpc/src/rpc_methods/get_transaction_receipt.rs [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+    C: HeaderBackend<B> + BlockBackend<B> + StorageProvider<B, BE> + 'static,
+    C: ProvideRuntimeApi<B>,
+    C::Api: StarknetRuntimeApi<B> + ConvertTransactionRuntimeApi<B>,
+    G: GenesisProvider + Send + Sync + 'static,
+    H: HasherT + Send + Sync + 'static,
+{
+    let execution_infos = client
+		.client
+		.runtime_api()
+		.re_execute_transactions(previous_block_hash, vec![], transactions.clone())
```
Comment (jbcaron) [inline crates/client/rpc/src/rpc_methods/get_transaction_receipt.rs]: you can modify the parameters like this to retrieve only the information from the last transaction
```rust
re_re_execute_transactions(previous_block_hash, <previous_tx>, <last_tx>)
```
and remove `pop()`

File: crates/client/rpc/src/rpc_methods/get_transaction_receipt.rs [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+		})?
+		.map_err(|e| {
+			log::error!("Failed to reexecute the transactions: {e:?}");
+			StarknetRpcApiError::InternalServerError
+		})?
+		.map_err(|_| {
+			log::error!("One of the transaction failed during it's reexecution");
+			StarknetRpcApiError::InternalServerError
+		})?
+		.pop() // get only the last transaction execution info
```
Comment (jbcaron) [inline crates/client/rpc/src/rpc_methods/get_transaction_receipt.rs]: remove` pop()`
*in reference to the comment above*

**Review by jbcaron — APPROVED** [review]

**Pull request merged**